hitl-setup/fpganet.service

```diff
@@ -0,0 +1,27 @@
+# SPDX-FileCopyrightText: 2024 Google LLC
+#
+# SPDX-License-Identifier: CC0-1.0
+
+[Unit]
+Description=Set up bridge network with FPGAs
+DefaultDependencies=no
+Wants=network.target systemd-udev-settle.service
+After=local-fs.target network-pre.target systemd-sysctl.service systemd-modules-load.service systemd-udev-settle.service
+Before=network.target shutdown.target network-online.target
+Conflicts=shutdown.target
+
+[Service]
+Type=oneshot
+RemainAfterExit=true
+EnvironmentFile=/etc/fpganet
+ExecStart=/usr/sbin/ip link add name $BRIDGE_NAME type bridge
+ExecStart=/usr/sbin/ip link set dev $BRIDGE_NAME up
+ExecStart=/usr/sbin/ip address add $INTERFACE_IP dev $BRIDGE_NAME
+ExecStart=/usr/bin/sh -c 'for INTERFACE in $INTERFACES; do /usr/sbin/ip link set dev $INTERFACE up; done'
+ExecStart=/usr/bin/sh -c 'for INTERFACE in $INTERFACES; do /usr/sbin/ip link set dev $INTERFACE master $BRIDGE_NAME; done'
+
+ExecStop=/usr/sbin/ip link delete name $BRIDGE_NAME type bridge
+
+[Install]
+WantedBy=multi-user.target
+WantedBy=network-online.target
```

Comment:
```suggestion
WantedBy=multi-user.target
WantedBy=network-online.target
WantedBy=$whatever-the-github-actions-runner-is
```

?

Reply:
Do you want to prevent the runner from running if this script failed? Because `Wanted` is not strong enough for that.

If it's just about ordering, that's already taken care of by `Before=network.target`.

Reply:
> Do you want to prevent the runner from running if this script failed?

Yes.

How would you specify this?

Reply:
I would add this to the runner:
```
[Unit]
Requires=fpganet.service
After=fpganet.service
```

but if you want the runner to be able to run if you `systemctl disable fpganet.service`, you could add this to `fpganet.service`:
```
[Unit]
Before=fpganet.service

[Install]
RequiredBy=actions.runner.bittide-bittide-hardware.hoeve-1.service
```

But this seems weird. Why would it suddenly no longer require the service just because you `systemctl disable`d it? On the other hand, if you do that, you might have good reason to? I dunno :-).

Reply:
I was under the (possibly naive) impression that just doing:
```
systemctl restart fpganet.service
systemctl enable fpganet.service
[...]
systemctl add-wants isc-dhcp-server.service fpganet.service
systemctl restart isc-dhcp-server.service
systemctl enable isc-dhcp-server.service
```

would be good enough to make sure that first the `fpganet.service` gets started, and then `isc-dhcp-server.service`, somewhere during boot up of the machine.

If you want to do it "really properly" you should probably add a dependency between the runner and the dhcp server.
Just depending on the setting up of the bridge isn't strictly enough.

Reply:
I don't really understand what you mean, but I don't see any ordering in the commands you give. Those commands will certainly not start the `fpganet.service` before `isc-dhcp-server.service`.

> Note that requirement dependencies do not influence the order in which services are started or stopped. This has to be configured independently with the `After=` or `Before=` options. If unit foo.service pulls in unit bar.service as configured with `Wants=` and no ordering is configured with `After=` or `Before=`, then both units will be started simultaneously and without any delay between them if foo.service is activated.

Reply:
> If you want to do it "really properly" you should probably add a dependency between the runner and the dhcp server.

Doing that (if I understand what you mean correctly) means that during bootup, it will wait for the DHCP server to be started before launching the runners. But the DHCP server doesn't need to be available until well into a CI run, not for the runner to be able to start. So I think this is overkill and not really the systemd "we want boot to be quick and ordering dependencies to be few" philosophy. I think we needed `fpganet.service` to be well and truly before the runner because otherwise Docker would not be able to bind the bridge interface into its runtime; in other words, the runner truly does depend on `fpganet.service` to be available when it is starting.

But I may be misinterpreting, there are so many ways to interpret "add a dependency" in systemd land.

Reply:
>  I think we needed fpganet.service to be well and truly before the runner because otherwise Docker would not be able to bind the bridge interface into its runtime;

This was an earlier experiment, for the current version, the runner can start fine without the bridge. Obviously CI jobs that need the bridge will fail without it, but the runner starts fine and other jobs run fine.